include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -0,0 +1,371 @@
+              const Vector<Number> &src) const;
+
+  /**
+   * Implements a matrix-vector product with the underlying matrix as
+   * described in the main documentation of this class. Same as the other
+   * vmult function, but operating on plain pointers rather than a vector (no
+   * check of array bounds possible).
+   */
+  void vmult (Number *dst,
+              const Number *src) const;
+
+  /**
+   * Implements a matrix-vector product with the underlying matrix as
+   * described in the main documentation of this class.
+   */
+  void apply_inverse (Vector<Number> &dst,
+                      const Vector<Number> &src) const;
+
+  /**
+   * Implements a matrix-vector product with the underlying matrix as
+   * described in the main documentation of this class. Same as the other
+   * vmult function, but operating on plain pointers rather than a vector (no
+   * check of array bounds possible).
+   */
+  void apply_inverse (Number *dst,
+                      const Number *src) const;
+
+private:
+  /**
+   * A copy of the @p mass_matrix object passed to the reinit() method.
+   */
+  FullMatrix<Number> mass_matrix;
+
+  /**
+   * A copy of the @p derivative_matrix object passed to the reinit() method.
+   */
+  FullMatrix<Number> derivative_matrix;
+
+  /**
+   * A vector containing the generalized eigenvalues of A s = lambda B s.
+   */
+  AlignedVector<Number> eigenvalues;
+
+  /**
+   * The matrix containing the generalized eigenvectors.
+   */
+  Table<2,Number> eigenvectors;
+
+  /**
+   * An array for temporary data.
+   */
+  mutable AlignedVector<Number> tmp_array;
+};
+
+
+/*----------------------- Inline functions ----------------------------------*/
+
+#ifndef DOXYGEN
+
+
+template <int dim, typename Number, int size>
+inline
+TensorProductMatrix<dim,Number,size>::TensorProductMatrix()
+{}
+
+
+
+template <int dim, typename Number, int size>
+inline
+void
+TensorProductMatrix<dim,Number,size>
+::reinit(const FullMatrix<Number> &mass_matrix,
+         const FullMatrix<Number> &derivative_matrix)
+{
+  Assert(size == -1 ||
+         (size > 0 && static_cast<unsigned int>(size) == mass_matrix.m()),
+         ExcDimensionMismatch(size, mass_matrix.m()));
+  AssertDimension(mass_matrix.m(), mass_matrix.n());
+  AssertDimension(mass_matrix.m(), derivative_matrix.m());
+  AssertDimension(mass_matrix.m(), derivative_matrix.n());
+
+  this->mass_matrix = mass_matrix;
+  this->derivative_matrix = derivative_matrix;
+
+  std::vector<Vector<Number> > eigenvecs(mass_matrix.m());
+  LAPACKFullMatrix<Number> mass_copy(mass_matrix.m(), mass_matrix.n());
+  LAPACKFullMatrix<Number> deriv_copy(derivative_matrix.m(), derivative_matrix.n());
+  for (unsigned int i=0; i<mass_matrix.m(); ++i)
+    for (unsigned int j=0; j<mass_matrix.n(); ++j)
+      mass_copy(i,j) = mass_matrix(i,j);
+  for (unsigned int i=0; i<derivative_matrix.m(); ++i)
+    for (unsigned int j=0; j<derivative_matrix.n(); ++j)
+      deriv_copy(i,j) = derivative_matrix(i,j);
```

Comment:
Can't we just use `LAPACKFullMatrix<Number>::operator=(const FullMatrix &)` here?